SpecialContentStaging.php

```diff
@@ -18,138 +19,321 @@ function __construct() {
-				$this->copyPage( $this->pagePrefix, $page, $source, $target );
+			if ( $action === 'copy' ) {
+				$this->copyPage(
+					$this->pagePrefix,
+					WikiPage::newFromID( $page ),
+					$currStage,
+					$targetStage
+				);
+			}
+			if ( $action === 'archive' ) {
+				$this->archivePage( $page );
+			}
+			if ( $action === 'restore' ) {
+				$this->restorePage( $page );
 			}
 
 			$pages = array();
 			$allPages = $this->getPagesByPrefix( $this->pagePrefix );
-			foreach( $allPages as $page ) {
+			foreach ( $allPages as $page ) {
 				$stage = $this->getStage( $page->page_title );
 				$title = $this->getTitleWithoutPrefixes( $page->page_title );
-				if( !array_key_exists( $title, $pages ) ) {
-					$pages[$title] = $this->stages;
+
+				$wikiPage = WikiPage::newFromID( $page->page_id );
+
+				if ( $this->shouldPageShow( $wikiPage, $showArchived ) ) {
+					if ( !array_key_exists( $title, $pages ) ) {
+						$pages[$title] = $this->stages;
+					}
+					$pages[$title][$stage] = $wikiPage;
 				}
-				$pages[$title][$stage] = WikiPage::newFromID( $page->page_id );
-				if ( $action === "stageall" && $source === $stage ) {
-					$pages[$title][$target] = $this->copyPage( $this->pagePrefix, $page->page_id, $source, $target );
+
+				if ( $action === "stageall" && $currStage === $stage ) {
+					$pages[$title][$targetStage] = $this->copyPage( $this->pagePrefix, $wikiPage, $currStage, $targetStage );
 				}
 			}
 
 			$resultTable = "{| class=\"wikitable sortable\" border=\"1\"\n";
 			$resultTable .= "|-\n";
-			$resultTable .= "! Titel\n";
-			$resultTable .= "! Test <html><br /><a href='?title=Special:ContentStaging&prefix=" . $this->pagePrefix . "&action=stageall&source=test&target=stage'>Stage all</a></html>\n";
-			$resultTable .= "! Stage <html><br /><a href='?title=Special:ContentStaging&prefix=" . $this->pagePrefix . "&action=stageall&source=stage&target=production'>Stage all</a></html>\n";
-			$resultTable .= "! Produktion\n";
-			
-			foreach( $pages as $title => $stages ) {
+			$resultTable .= "! Title\n";
+			$resultTable .= "! Test <html><br /><a href='" . $baseUrl . "&action=stageall&source=test&target=stage'>Stage all</a></html>\n";
+			$resultTable .= "! Stage <html><br /><a href='" . $baseUrl . "&action=stageall&source=stage&target=production'>Stage all</a></html>\n";
+			$resultTable .= "! Production\n";
+			$resultTable .= "! Options\n";
+
+			foreach ( $pages as $title => $stages ) {
 				$resultTable .= "|-\n";
 				$resultTable .= "| [[" . $this->mwNamespace . $this->pagePrefix . "/test/" . $title . " | " . $title . "]]\n";
 
-				foreach( array_keys( $this->stages ) as $stage ) {
-					if( $stages[$stage] !== 0 ) {
-						$target = "";
-						$source = $stage;
-						$keys = array_keys( $stages );
-						$element = array_search( $source, $keys );
-						if ( $stage !== "production" ) $target = $keys[$element + 1];
+				foreach ( array_keys( $this->stages ) as $stage ) {
+					if ( $stages[$stage] !== 0 ) {
```

Comment:
That's what I thought. I also had other numbers than zeroes before, not knowing what difference those numbers make. Actually I still don't know.
But if it is actually intended to have either page object or zero, maybe switch to null instead of zero?